cmd/cmd.go

```diff
@@ -70,3 +80,394 @@ func validateUserConfig(cfg *viper.Viper) error {
+// Non-200 responses and 0 Content-Length responses are swallowed, returning nil.
+func (d *download) requestFile(filename string) (*http.Response, error) {
+	parsedURL, err := netURL.ParseRequestURI(
+		fmt.Sprintf("%s%s", d.Solution.FileDownloadBaseURL, filename))
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := api.NewClient(d.params.token, d.params.apibaseurl)
+	req, err := client.NewRequest("GET", parsedURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		// TODO: deal with it
+		return nil, nil
+	}
+	// Don't bother with empty files.
+	if res.Header.Get("Content-Length") == "0" {
+		return nil, nil
+	}
+
+	return res, nil
+}
+
+func (d *download) metadata() ws.ExerciseMetadata {
+	return ws.ExerciseMetadata{
+		AutoApprove: d.Solution.Exercise.AutoApprove,
+		Track:       d.Solution.Exercise.Track.ID,
+		Team:        d.Solution.Team.Slug,
+		Exercise:    d.Solution.Exercise.ID,
+		ID:          d.Solution.ID,
+		URL:         d.Solution.URL,
+		Handle:      d.Solution.User.Handle,
+		IsRequester: d.Solution.User.IsRequester,
+	}
+}
+
+func (d *download) exercise() ws.Exercise {
+	return ws.Exercise{
+		Root:  d.solutionRoot(),
+		Track: d.Solution.Exercise.Track.ID,
+		Slug:  d.Solution.Exercise.ID,
+	}
+}
+
+// solutionRoot builds the root path based on the solution
+// being part of a team and/or owned by another user.
+func (d *download) solutionRoot() string {
+	root := d.params.workspace
+
+	if d.isTeamSolution() {
+		root = filepath.Join(root, "teams", d.Solution.Team.Slug)
+	}
+	if d.solutionBelongsToOtherUser() {
+		root = filepath.Join(root, "users", d.Solution.User.Handle)
+	}
+	return root
+}
+
+func (d *download) isTeamSolution() bool {
+	return d.Solution.Team.Slug != ""
+}
+
+func (d *download) solutionBelongsToOtherUser() bool {
+	return !d.Solution.User.IsRequester
+}
+
+// validate validates the presence of an ID and checks for an error message.
+func (d *download) validate() error {
+	if d.Solution.ID == "" {
+		return errors.New("download missing an ID")
+	}
+	if d.Error.Message != "" {
+		return errors.New(d.Error.Message)
+	}
+	return nil
+}
+
+// downloadWriter writes download contents to the workspace.
+type downloadWriter struct {
+	*download
+}
+
+// writeMetadata writes the exercise metadata.
+func (d downloadWriter) writeMetadata() error {
+	metadata := d.metadata()
+	return metadata.Write(d.destination())
+}
+
+// writeSolutionFiles attempts to write each exercise file that is part of the downloaded Solution.
+// An HTTP request is made using each filename and failed responses are swallowed.
+// All successful file responses are written except when 0 Content-Length.
+func (d downloadWriter) writeSolutionFiles() error {
+	if d.params.fromExercise {
+		return errors.New("existing exercise files should not be overwritten")
```

Comment:
This probably isn't quite what this error should say. But if `doctor` shouldn't be allowed to write exercise files, it seems like worth throwing an error here.